src/libraries/System.Security.Principal.Windows/tests/WindowsIdentityImpersonatedTests.netcoreapp.cs

```diff
@@ -0,0 +1,178 @@
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.ComponentModel;
+using System.Runtime.InteropServices;
+using System.Security.Cryptography;
+using System.Security.Principal;
+using System.Threading.Tasks;
+using Microsoft.Win32.SafeHandles;
+using Xunit;
+
+// On nano server netapi32.dll is not present
+// we'll skip all tests on that platform
+public class WindowsIdentityImpersonatedTests : IClassFixture<WindowsIdentityFixture>
+{
+    private readonly WindowsIdentityFixture _fixture;
+
+    public WindowsIdentityImpersonatedTests(WindowsIdentityFixture windowsIdentityFixture)
+    {
+        _fixture = windowsIdentityFixture;
+
+        Assert.False(_fixture.TestAccount.AccountTokenHandle.IsInvalid);
+        Assert.False(string.IsNullOrEmpty(_fixture.TestAccount.AccountName));
+    }
+
+    [ConditionalFact(typeof(PlatformDetection), nameof(PlatformDetection.IsNotWindowsNanoServer))]
+    [OuterLoop]
+    public async Task RunImpersonatedAsync_TaskAndTaskOfT()
+    {
+        WindowsIdentity currentWindowsIdentity = WindowsIdentity.GetCurrent();
+
+        await WindowsIdentity.RunImpersonatedAsync(_fixture.TestAccount.AccountTokenHandle, async () =>
+        {
+            Asserts(currentWindowsIdentity);
+            await Task.Delay(100);
+            Asserts(currentWindowsIdentity);
+        });
+
+        Assert.Equal(WindowsIdentity.GetCurrent().Name, currentWindowsIdentity.Name);
+
+        int result = await WindowsIdentity.RunImpersonatedAsync(_fixture.TestAccount.AccountTokenHandle, async () =>
+        {
+            Asserts(currentWindowsIdentity);
+            await Task.Delay(100);
+            Asserts(currentWindowsIdentity);
+            return 42;
+        });
+
+        Assert.Equal(42, result);
+        Assert.Equal(WindowsIdentity.GetCurrent().Name, currentWindowsIdentity.Name);
+
+        // Assertions
+        void Asserts(WindowsIdentity currentWindowsIdentity)
+        {
+            Assert.Equal(_fixture.TestAccount.AccountName, WindowsIdentity.GetCurrent().Name);
+            Assert.NotEqual(currentWindowsIdentity.Name, WindowsIdentity.GetCurrent().Name);
+        }
+    }
+}
+
+public class WindowsIdentityFixture : IDisposable
+{
+    public WindowsTestAccount TestAccount { get; private set; }
+
+    public WindowsIdentityFixture()
+    {
+        TestAccount = new WindowsTestAccount("CorFxTstWiIde01kiu");
+        TestAccount.Create();
+    }
+
+    public void Dispose()
+    {
+        TestAccount.Dispose();
+    }
+}
+
+public sealed class WindowsTestAccount : IDisposable
+{
+    private readonly string _userName;
+    private SafeAccessTokenHandle _accountTokenHandle;
+    public SafeAccessTokenHandle AccountTokenHandle => _accountTokenHandle;
+    public string AccountName
+    {
+        get
+        {
+            // We should not use System.Security.Principal.Windows classes that we'are testing.
```

Comment:
Why not?  We're validating how impersonation flows, not the functionality of these types.  Other functions elsewhere in the tests should be validating the behavior of the methods we'd rely on here, and then we can trust that they're correct and use them here rather than duplicating all that functionality in tests.

Reply:
Because get name run impersonation code flow that we're testing https://github.com/dotnet/runtime/blob/master/src/libraries/System.Security.Principal.Windows/src/System/Security/Principal/WindowsIdentity.cs#L554 a bug in retrive name could lead to false positive or complex fails to investigate, don't you agree?

Reply:
> a bug in retrive name could lead to false positive or complex fails to investigate, don't you agree?

Yes, but shouldn't such a bug be caught by tests for Name?

Reply:
Should be, I don't see deep test on it https://github.com/dotnet/runtime/blob/4f9ae42d861fcb4be2fcd5d3d55d5f227d30e723/src/libraries/System.Security.Principal.Windows/tests/WindowsIdentityTests.cs#L88 and the token used to clone and check is got by `WindowsIdentity.GetCurrent()`.

BTW I'll use `WindowsIdentity`, lack of deeper tests is another issue.